### PR TITLE
feat(js): adds elgg/popup AMD module

### PR DESCRIPTION
**Pull request by hypeJunction**
Converts JS popup to an AMD module subsequently allowing plugins to
invoke and close popup programmically.
Popups can now be bound with .elgg-popup class in addition to rel=popup.
Popup positioning details can be passed with data- attributes of the trigger.

### CODE REVIEW

**Comment by hypeJunction** [comment]
Let's see if you think this will work in 2.x. I think it's BC, though I refactored the code for opening and closing. Once this is in:

- [ ] Move user-hover menu logic to use the `elgg/popup` module to get rid of code duplication
- [ ] In 3.0, do not append the module to `body` - breaking change that will probably require z-index changes.

File: docs/guides/javascript.rst [hunk 1]
```diff
@@ -434,6 +434,59 @@ The ``elgg/spinner`` module can be used to create an Ajax loading indicator fixe
...
+      'text' => 'Open popup',
+      'rel' => 'popup',
+   ]);
+
+   // Button with custom positioning of the popup
+   echo elgg_format_element('a', [
+      'class' => 'elgg-popup elgg-button elgg-button-dropdown',
+      'href' => '#popup-element',
+      'text' => 'Open popup above',
+      'data-my' => 'center bottom',
```
Comment (hypeJunction) [inline docs/guides/javascript.rst]: Should we make it `data-position` with json?

**Comment by hypeJunction** [comment]
Screenshot from the theme sandbox
popup-button

File: docs/guides/javascript.rst (hunk 1 above)
Comment (mrclay) [inline docs/guides/javascript.rst]: :+1:

File: js/lib/ui.js [hunk 2]
```diff
@@ -18,7 +18,9 @@ elgg.ui.init = function () {
 
 	$(document).on('click', '[rel=toggle]', elgg.ui.toggles);
 
-	$(document).on('click', '[rel=popup]', elgg.ui.popupOpen);
```
Comment (mrclay) [inline js/lib/ui.js]: Old code handles HTML added by Ajax. For BC we should do that as well.

Comment (hypeJunction) [inline js/lib/ui.js]: This works with AJAX. The bindings are proxied.

**Comment by hypeJunction** [comment]
Improved the guides and added some inline docs.
Now uses `data-position`, and also accepts `data-href` if `href` is not present.
User hover menu now uses the `elgg/popup` module.
Added jQuery events we can use to populate the popup dynamically.

@juho-jaakkola Can you take a look, I had notifier in mind when working on this.

File: js/lib/ui.js [hunk 3]
```diff
@@ -100,87 +105,30 @@ elgg.ui.toggles = function(event) {
  *
  * @param {Object} event
  * @return void
+ * @deprecated 2.1
```
Comment (mrclay) [inline js/lib/ui.js]: 2.2

File: views/default/elgg/popup.js [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+			$(document).on('click', function (e) {
+				var $eventTargets = $(e.target).parents().andSelf();
+				if ($eventTargets.is('[data-popup]')) {
+					return;
+				}
+				popup.close();
+			});
+
+			// Adding this magic so that popups with fixed position stick to their parent element
+			$(window).on('scroll resize', function () {
```
Comment (mrclay) [inline views/default/elgg/popup.js]: Should we throttle this event by 100ms? Or can we detect fixed position to make sure we have to do this?

Comment (mrclay) [inline views/default/elgg/popup.js]: Is this required for the AMD migration or is this a bugfix we could move to another PR?

Comment (hypeJunction) [inline views/default/elgg/popup.js]: I believe currently it remains in the viewport when you scroll, so it's a UX bug. I will see if I can migrate with minimal changes and then fluff it up in additional PRs

**Comment by hypeJunction** [comment]
Replaced by #9599
